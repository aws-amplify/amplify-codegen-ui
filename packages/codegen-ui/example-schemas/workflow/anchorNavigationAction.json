{
  "componentType": "Button",
  "name": "NavigateButton",
  "properties": {
    "children": {
      "value": "Go to About section"
    }
  },
  "events": {
    "onClick": {
      "action": "Amplify.Navigation",
      "parameters": {
        "type": {
          "value": "anchor"
        },
        "about": {
          "value": "#about"
        }
      }
    }
  },
  "schemaVersion": "1.0"
}
