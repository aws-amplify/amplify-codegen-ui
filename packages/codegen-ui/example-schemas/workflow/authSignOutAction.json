{
  "componentType": "Button",
  "name": "SignOutButton",
  "properties": {
    "label": {
      "value": "Sign out"
    }
  },
  "events": {
    "click": {
      "action": "Amplify.AuthSignOut",
      "parameters": {
        "global": {
          "value": false
        }
      }
    }
  }
}
