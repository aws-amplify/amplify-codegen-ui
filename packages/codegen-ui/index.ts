/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
export * from './lib/template-renderer';
export * from './lib/frontend-manager-template-renderer';

export * from './lib/component-renderer-base';
export * from './lib/component-with-children-renderer-base';
export * from './lib/render-component-response';
export * from './lib/framework-output-manager';
export * from './lib/template-renderer-factory';

export * from './lib/renderer-helper';
export * from './lib/validation-helper';
export * from './lib/framework-output-config';
export * from './lib/framework-render-config';
export * from './lib/frontend-manager-node';

export * from './lib/types';
export * from './lib/errors';
