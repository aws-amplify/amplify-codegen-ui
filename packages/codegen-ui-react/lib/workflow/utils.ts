/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { ComponentMetadata, InvalidInputError } from '@aws-amplify/codegen-ui';
import Primitive, { PrimitiveChildrenPropMapping } from '../primitive';

export const getChildPropMappingForComponentName = (
  componentMetadata: ComponentMetadata,
  componentName: string,
): string | undefined => {
  const referencedComponentType = componentMetadata.componentNameToTypeMap[componentName];
  if (!referencedComponentType) {
    throw new InvalidInputError(
      `Invalid definition, found reference to component name ${componentName} which wasn't found in the schema.`,
    );
  }
  return PrimitiveChildrenPropMapping[Primitive[referencedComponentType as Primitive]];
};
