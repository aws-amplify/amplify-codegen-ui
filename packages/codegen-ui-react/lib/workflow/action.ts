/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import ts, { Statement, factory, ObjectLiteralExpression, Expression, PropertyAssignment } from 'typescript';
import {
  InvalidInputError,
  StudioComponent,
  StudioComponentChild,
  ActionStudioComponentEvent,
  StudioComponentProperty,
  MutationAction,
  ComponentMetadata,
  DataStoreCreateItemAction,
  DataStoreUpdateItemAction,
  DataStoreDeleteItemAction,
} from '@aws-amplify/codegen-ui';
import { isActionEvent, propertyToExpression, getSetStateName, sanitizeName } from '../react-component-render-helper';
import { ImportCollection, ImportSource, ImportValue } from '../imports';
import { getChildPropMappingForComponentName } from './utils';
import { DataApiKind } from '../react-render-config';
import { ActionType, getGraphqlCallExpression } from '../utils/graphql';

enum Action {
  'Amplify.Navigation' = 'Amplify.Navigation',
  'Amplify.DataStoreCreateItemAction' = 'Amplify.DataStoreCreateItemAction',
  'Amplify.DataStoreUpdateItemAction' = 'Amplify.DataStoreUpdateItemAction',
  'Amplify.DataStoreDeleteItemAction' = 'Amplify.DataStoreDeleteItemAction',
  'Amplify.AuthSignOut' = 'Amplify.AuthSignOut',
  'Amplify.Mutation' = 'Amplify.Mutation',
}

type DataAction = DataStoreCreateItemAction | DataStoreUpdateItemAction | DataStoreDeleteItemAction;

export default Action;

export const ActionNameMapping: Partial<Record<Action, ImportValue>> = {
  [Action['Amplify.Navigation']]: ImportValue.USE_NAVIGATE_ACTION,
  [Action['Amplify.DataStoreCreateItemAction']]: ImportValue.USE_DATA_STORE_CREATE_ACTION,
  [Action['Amplify.DataStoreUpdateItemAction']]: ImportValue.USE_DATA_STORE_UPDATE_ACTION,
  [Action['Amplify.DataStoreDeleteItemAction']]: ImportValue.USE_DATA_STORE_DELETE_ACTION,
  [Action['Amplify.AuthSignOut']]: ImportValue.USE_AUTH_SIGN_OUT_ACTION,
  [Action['Amplify.Mutation']]: ImportValue.USE_STATE_MUTATION_ACTION,
};

const DataStoreActions = new Set([
  Action['Amplify.DataStoreCreateItemAction'],
  Action['Amplify.DataStoreDeleteItemAction'],
  Action['Amplify.DataStoreUpdateItemAction'],
]);

function isMutationAction(action: ActionStudioComponentEvent): action is MutationAction {
  return (action.action as Action) === Action['Amplify.Mutation'];
}

function isDataAction(action: ActionStudioComponentEvent): action is DataAction {
  return DataStoreActions.has(action.action as Action);
}

export function getActionHookImportValue(action: string): ImportValue {
  const actionName = ActionNameMapping[Action[action as Action]];
  if (actionName === undefined) {
    throw new InvalidInputError(`${action} is not a valid action.`);
  }
  return actionName;
}

function getGraphqlMutationFromAction(action: Action): ActionType {
  switch (action) {
    case Action['Amplify.DataStoreCreateItemAction']:
      return ActionType.CREATE;
    case Action['Amplify.DataStoreUpdateItemAction']:
      return ActionType.UPDATE;
    case Action['Amplify.DataStoreDeleteItemAction']:
      return ActionType.DELETE;
    default:
      throw new InvalidInputError(`Action ${action} has no corresponding GraphQL operation`);
  }
}

export function getComponentActions(component: StudioComponent | StudioComponentChild): {
  action: ActionStudioComponentEvent;
  identifier: string;
}[] {
  const actions = [];
  if (component.events) {
    actions.push(
      ...Object.entries(component.events)
        // need 'action is ...' so that typescript will know all items after filter are ActionStudioComponentEvent type
        .filter((action): action is [string, ActionStudioComponentEvent] => isActionEvent(action[1]))
        .map(([event, action]) => ({ action, identifier: getActionIdentifier(component.name, event) })),
    );
  }
  if (component.children) {
    actions.push(...component.children.map(getComponentActions).flat());
  }
  return actions;
}

export function getActionIdentifier(componentName: string, event: string) {
  const name = sanitizeName(componentName);
  return [name.charAt(0).toLowerCase() + name.slice(1), event.charAt(0).toUpperCase() + event.slice(1)].join('');
}

export function buildUseActionStatement(
  componentMetadata: ComponentMetadata,
  action: ActionStudioComponentEvent,
  identifier: string,
  importCollection: ImportCollection,
  apiKind: DataApiKind = 'DataStore',
  renderConfigDependencies?: { [key: string]: string },
): Statement {
  if (isMutationAction(action)) {
    return buildMutationActionStatement(componentMetadata, action, identifier);
  }

  if (isDataAction(action) && apiKind === 'GraphQL') {
    return buildGraphqlCallback(componentMetadata, action, identifier, importCollection, renderConfigDependencies);
  }

  const actionHookImportValue = getActionHookImportValue(action.action);
  importCollection.addMappedImport(actionHookImportValue);
  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(identifier),
          undefined,
          undefined,
          factory.createCallExpression(
            factory.createIdentifier(actionHookImportValue),
            undefined,
            buildActionArguments(componentMetadata, action, importCollection),
          ),
        ),
      ],
      ts.NodeFlags.Const,
    ),
  );
}

export function buildGraphqlCallback(
  componentMetadata: ComponentMetadata,
  action: DataAction,
  identifier: string,
  importCollection: ImportCollection,
  renderConfigDependencies?: { [key: string]: string },
) {
  const inputKeys = [];

  if ('id' in action.parameters && action.parameters.id) {
    inputKeys.push(
      factory.createPropertyAssignment(
        'id',
        getActionParameterValue(componentMetadata, 'id', action.parameters.id, importCollection),
      ),
    );
  }
  if ('fields' in action.parameters && action.parameters.fields) {
    inputKeys.unshift(...assignFieldProperties(componentMetadata, action.parameters.fields, importCollection));
  }

  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(identifier),
          undefined,
          undefined,
          factory.createArrowFunction(
            [factory.createToken(ts.SyntaxKind.AsyncKeyword)],
            undefined,
            [],
            undefined,
            factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
            factory.createBlock(
              [
                factory.createExpressionStatement(
                  factory.createAwaitExpression(
                    getGraphqlCallExpression(
                      getGraphqlMutationFromAction(action.action as Action),
                      action.parameters.model,
                      importCollection,
                      { inputs: inputKeys },
                      undefined,
                      renderConfigDependencies,
                    ),
                  ),
                ),
              ],
              true,
            ),
          ),
        ),
      ],
      ts.NodeFlags.Const,
    ),
  );
}

export function buildMutationActionStatement(
  componentMetadata: ComponentMetadata,
  action: MutationAction,
  identifier: string,
) {
  const { componentName, property } = action.parameters.state;
  const childrenPropMapping = getChildPropMappingForComponentName(componentMetadata, componentName);
  const stateReference =
    childrenPropMapping !== undefined && property === childrenPropMapping
      ? { ...action.parameters.state, property: 'children' }
      : action.parameters.state;

  const setState = getSetStateName(stateReference);

  return factory.createVariableStatement(
    undefined,
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier(identifier),
          undefined,
          undefined,
          factory.createArrowFunction(
            undefined,
            undefined,
            [],
            undefined,
            factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
            factory.createBlock(
              [
                factory.createExpressionStatement(
                  factory.createCallExpression(factory.createIdentifier(setState), undefined, [
                    propertyToExpression(componentMetadata, action.parameters.state.set),
                  ]),
                ),
              ],
              true,
            ),
          ),
        ),
      ],
      ts.NodeFlags.Const,
    ),
  );
}

/* Transform the action parameters field to literal
 *
 * model and fields are special cases. All other fields are StudioComponentProperty
 */
export function buildActionArguments(
  componentMetadata: ComponentMetadata,
  action: ActionStudioComponentEvent,
  importCollection: ImportCollection,
): ObjectLiteralExpression[] | undefined {
  if (action.parameters) {
    return [buildActionArgument(componentMetadata, action, importCollection)];
  }

  return undefined;
}

function buildActionArgument(
  componentMetadata: ComponentMetadata,
  action: ActionStudioComponentEvent,
  importCollection: ImportCollection,
): ObjectLiteralExpression {
  const properties = Object.entries(action.parameters).map(([key, value]) =>
    factory.createPropertyAssignment(
      factory.createIdentifier(key),
      getActionParameterValue(componentMetadata, key, value, importCollection),
    ),
  );

  if (DataStoreActions.has(action.action as Action)) {
    addSchemaToArguments(properties, importCollection);
  }
  return factory.createObjectLiteralExpression(properties, false);
}

export function addSchemaToArguments(properties: ts.PropertyAssignment[], importCollection: ImportCollection) {
  const SCHEMA = 'schema';
  properties.push(factory.createPropertyAssignment(factory.createIdentifier(SCHEMA), factory.createIdentifier(SCHEMA)));
  importCollection.addImport(ImportSource.LOCAL_SCHEMA, SCHEMA);
}

function assignFieldProperties(
  componentMetadata: ComponentMetadata,
  fieldsValue: StudioComponentProperty | { [key: string]: StudioComponentProperty } | string,
  importCollection: ImportCollection,
): PropertyAssignment[] {
  return Object.entries(fieldsValue).map(([nestedKey, nestedValue]) =>
    factory.createPropertyAssignment(
      factory.createIdentifier(nestedKey),
      getActionParameterValue(componentMetadata, nestedKey, nestedValue, importCollection),
    ),
  );
}

export function getActionParameterValue(
  componentMetadata: ComponentMetadata,
  key: string,
  value: StudioComponentProperty | { [key: string]: StudioComponentProperty } | string,
  importCollection: ImportCollection,
): Expression {
  if (key === 'model') {
    const modelName = importCollection.addModelImport(value as string);
    return factory.createIdentifier(modelName);
  }
  if (key === 'fields') {
    return factory.createObjectLiteralExpression(
      assignFieldProperties(componentMetadata, value, importCollection),
      false,
    );
  }
  return propertyToExpression(componentMetadata, value as StudioComponentProperty);
}
