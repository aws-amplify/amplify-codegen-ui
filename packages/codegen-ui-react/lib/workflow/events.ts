/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import {
  StudioGenericEvent,
  StudioComponentEvent,
  BoundStudioComponentEvent,
  ActionStudioComponentEvent,
} from '@aws-amplify/codegen-ui';

import { factory, JsxAttribute, SyntaxKind } from 'typescript';

import { getActionIdentifier } from './action';
import { isBoundEvent, isActionEvent } from '../react-component-render-helper';

export function buildOpeningElementEvents(
  event: StudioComponentEvent,
  name: string,
  componentName: string | undefined,
): JsxAttribute {
  if (isBoundEvent(event)) {
    return buildBindingEvent(event, name);
  }
  if (isActionEvent(event)) {
    return buildActionEvent(event, name, componentName);
  }

  return factory.createJsxAttribute(factory.createIdentifier(name), undefined);
}

export function buildBindingEvent(event: BoundStudioComponentEvent, eventName: string): JsxAttribute {
  const expr = factory.createIdentifier(event.bindingEvent);
  return factory.createJsxAttribute(
    factory.createIdentifier(mapGenericEventToReact(eventName as StudioGenericEvent)),
    factory.createJsxExpression(undefined, expr),
  );
}

export function buildActionEvent(
  event: ActionStudioComponentEvent,
  eventName: string,
  componentName: string | undefined,
): JsxAttribute {
  return factory.createJsxAttribute(
    factory.createIdentifier(mapGenericEventToReact(eventName as StudioGenericEvent)),
    factory.createJsxExpression(
      undefined,
      factory.createArrowFunction(
        undefined,
        undefined,
        [],
        undefined,
        factory.createToken(SyntaxKind.EqualsGreaterThanToken),
        factory.createBlock(
          [
            factory.createExpressionStatement(
              factory.createCallExpression(
                factory.createIdentifier(getActionIdentifier(componentName, eventName)),
                undefined,
                [],
              ),
            ),
          ],
          false,
        ),
      ),
    ),
  );
}

/*
 * Temporary hardcoded mapping of generic to react events, long-term this will be exported by amplify-ui.
 */
const genericEventToReactEventMapping = {
  [StudioGenericEvent.click]: 'onClick',
  [StudioGenericEvent.doubleclick]: 'onDoubleClick',
  [StudioGenericEvent.mousedown]: 'onMouseDown',
  [StudioGenericEvent.mouseenter]: 'onMouseEnter',
  [StudioGenericEvent.mouseleave]: 'onMouseLeave',
  [StudioGenericEvent.mousemove]: 'onMouseMove',
  [StudioGenericEvent.mouseout]: 'onMouseOut',
  [StudioGenericEvent.mouseover]: 'onMouseOver',
  [StudioGenericEvent.mouseup]: 'onMouseUp',
  [StudioGenericEvent.change]: 'onChange',
  [StudioGenericEvent.input]: 'onInput',
  [StudioGenericEvent.focus]: 'onFocus',
  [StudioGenericEvent.blur]: 'onBlur',
  [StudioGenericEvent.keydown]: 'onKeyDown',
  [StudioGenericEvent.keypress]: 'onKeyPress',
  [StudioGenericEvent.keyup]: 'onKeyUp',
};

export function mapGenericEventToReact(genericEventBinding: StudioGenericEvent): string {
  const reactEventMapping = genericEventToReactEventMapping[genericEventBinding];
  if (!reactEventMapping) {
    throw new Error(`${genericEventBinding} is not a possible event.`);
  }
  return reactEventMapping;
}
