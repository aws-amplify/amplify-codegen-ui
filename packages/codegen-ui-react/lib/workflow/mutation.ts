/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import ts, { Statement, factory, JsxAttribute, JsxExpression } from 'typescript';
import {
  StudioComponent,
  StudioComponentChild,
  StateStudioComponentProperty,
  ActionStudioComponentEvent,
  StudioGenericEvent,
  StudioComponentProperty,
  ComponentMetadata,
  getComponentFromComponentTree,
  StateReferenceMetadata,
} from '@aws-amplify/codegen-ui';
import {
  isStateProperty,
  isSetStateParameter,
  propertyToExpression,
  getStateName,
  getSetStateName,
} from '../react-component-render-helper';
import { ImportCollection, ImportValue } from '../imports';
import { mapGenericEventToReact } from './events';
import { getChildPropMappingForComponentName } from './utils';
import {
  Primitive,
  PrimitivesWithChangeEvent,
  PrimitiveLevelPropConfiguration,
  PrimitiveDefaultPropertyValue,
} from '../primitive';
import { DataApiKind } from '../react-render-config';

type EventHandlerBuilder = (setStateName: string, stateName: string) => JsxExpression;

const genericEventToReactEventImplementationOverrides: PrimitiveLevelPropConfiguration<EventHandlerBuilder> = {
  [Primitive.StepperField]: { [StudioGenericEvent.onChange]: numericValueCallback },
  [Primitive.SliderField]: { [StudioGenericEvent.onChange]: numericValueCallback },
  [Primitive.CheckboxField]: { [StudioGenericEvent.onChange]: buildSyntheticEventTargetCallback('checked') },
  [Primitive.SwitchField]: { [StudioGenericEvent.onChange]: toggleBooleanStateCallback },
};

const PrimitiveDefaultValuePropMapping: PrimitiveLevelPropConfiguration<string> = new Proxy(
  {
    [Primitive.SwitchField]: { isChecked: 'defaultChecked' },
  },
  {
    get(target, name, ...args) {
      return name in target ? Reflect.get(target, name, ...args) : { value: 'defaultValue' };
    },
  },
);

export function buildUseEffectStatements(
  component: StudioComponent,
  componentMetadata: ComponentMetadata,
): Statement[] {
  return componentMetadata.stateReferences
    .filter(({ dataDependencies }) => dataDependencies.length > 0)
    .map((stateReference) => {
      const { reference, dataDependencies } = stateReference;
      const dependencyConditions = dataDependencies.flatMap((dataDependency) => [
        // prevents components from changing to an uncontrolled state
        factory.createBinaryExpression(
          getStateInitialValue(component, componentMetadata, { reference, dataDependencies: [] }),
          factory.createToken(ts.SyntaxKind.ExclamationEqualsEqualsToken),
          factory.createIdentifier('undefined'),
        ),
        factory.createBinaryExpression(
          factory.createIdentifier(dataDependency),
          factory.createToken(ts.SyntaxKind.ExclamationEqualsEqualsToken),
          factory.createIdentifier('undefined'),
        ),
      ]);
      const conditions = [
        ...dependencyConditions,
        factory.createBinaryExpression(
          factory.createIdentifier(getStateName(reference)),
          factory.createToken(ts.SyntaxKind.EqualsEqualsEqualsToken),
          getStateInitialValue(component, componentMetadata, stateReference),
        ),
      ];
      const binaryExpression = conditions.reduce((expression, currentCondition) => {
        return factory.createBinaryExpression(
          factory.createParenthesizedExpression(currentCondition),
          factory.createToken(ts.SyntaxKind.AmpersandAmpersandToken),
          expression,
        );
      });
      return factory.createExpressionStatement(
        factory.createCallExpression(factory.createIdentifier('useEffect'), undefined, [
          factory.createArrowFunction(
            undefined,
            undefined,
            [],
            undefined,
            factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
            factory.createBlock([
              factory.createIfStatement(
                binaryExpression,
                factory.createExpressionStatement(
                  factory.createCallExpression(factory.createIdentifier(getSetStateName(reference)), undefined, [
                    getStateInitialValue(component, componentMetadata, { reference, dataDependencies: [] }),
                  ]),
                ),
              ),
            ]),
          ),
          factory.createArrayLiteralExpression(dataDependencies.map(factory.createIdentifier), false),
        ]),
      );
    });
}

export function mapSyntheticStateReferences(componentMetadata: ComponentMetadata) {
  return componentMetadata.stateReferences.map((stateReference) => {
    const {
      reference: { componentName, property },
    } = stateReference;
    const childrenPropMapping = getChildPropMappingForComponentName(componentMetadata, componentName);
    if (childrenPropMapping !== undefined && property === childrenPropMapping) {
      return { ...stateReference, reference: { ...stateReference.reference, property: 'children' } };
    }
    return stateReference;
  });
}

export function getActionStateParameters(action: ActionStudioComponentEvent): StateStudioComponentProperty[] {
  if (action.parameters) {
    return Object.entries(action.parameters)
      .filter(([key]) => key !== 'model')
      .flatMap(([key, parameter]) => {
        if (key === 'fields' || key === 'attributes') {
          return Object.values(parameter);
        }
        if (isSetStateParameter(parameter)) {
          return [parameter, parameter.set];
        }
        return parameter;
      })
      .filter((parameter) => isStateProperty(parameter) || isSetStateParameter(parameter));
  }
  return [];
}

function buildSyntheticEventTargetCallback(targetFieldName: string): (setStateName: string) => JsxExpression {
  return (setStateName: string) => {
    return factory.createJsxExpression(
      undefined,
      factory.createArrowFunction(
        undefined,
        undefined,
        [
          factory.createParameterDeclaration(
            undefined,
            undefined,
            undefined,
            factory.createIdentifier('event'),
            undefined,
            factory.createTypeReferenceNode(factory.createIdentifier('SyntheticEvent'), undefined),
            undefined,
          ),
        ],
        undefined,
        factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
        factory.createBlock(
          [
            factory.createExpressionStatement(
              factory.createCallExpression(factory.createIdentifier(setStateName), undefined, [
                factory.createPropertyAccessExpression(
                  factory.createPropertyAccessExpression(
                    factory.createIdentifier('event'),
                    factory.createIdentifier('target'),
                  ),
                  factory.createIdentifier(targetFieldName),
                ),
              ]),
            ),
          ],
          false,
        ),
      ),
    );
  };
}

function numericValueCallback(setStateName: string): JsxExpression {
  return factory.createJsxExpression(
    undefined,
    factory.createArrowFunction(
      undefined,
      undefined,
      [
        factory.createParameterDeclaration(
          undefined,
          undefined,
          undefined,
          factory.createIdentifier('value'),
          undefined,
          factory.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword),
          undefined,
        ),
      ],
      undefined,
      factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
      factory.createCallExpression(factory.createIdentifier(setStateName), undefined, [
        factory.createIdentifier('value'),
      ]),
    ),
  );
}

function toggleBooleanStateCallback(setStateName: string, stateName: string): JsxExpression {
  return factory.createJsxExpression(
    undefined,
    factory.createArrowFunction(
      undefined,
      undefined,
      [],
      undefined,
      factory.createToken(ts.SyntaxKind.EqualsGreaterThanToken),
      factory.createCallExpression(factory.createIdentifier(setStateName), undefined, [
        factory.createPrefixUnaryExpression(ts.SyntaxKind.ExclamationToken, factory.createIdentifier(stateName)),
      ]),
    ),
  );
}

export function buildOpeningElementControlEvents(
  componentType: string,
  setStateName: string,
  stateName: string,
  event: string,
): JsxAttribute {
  const implementationOverrides = genericEventToReactEventImplementationOverrides[componentType];
  const controlEventBuilder =
    implementationOverrides && implementationOverrides[event]
      ? implementationOverrides[event]
      : buildSyntheticEventTargetCallback('value');
  return factory.createJsxAttribute(
    factory.createIdentifier(mapGenericEventToReact(componentType as Primitive, event as StudioGenericEvent)),
    controlEventBuilder(setStateName, stateName),
  );
}

/**
 * Dedupes state references by componentName, property, and dataDependencies, returning a consolidate
 * list, stripping the `set` property from them. We do this by serializing to json,
 * collecting in a set, then deserializing.
 */
function dedupeStateReferences(stateReferences: StateReferenceMetadata[]): StateReferenceMetadata[] {
  const dedupedSerializedRefs = [
    ...new Set(
      stateReferences.map((stateReference) => {
        const {
          reference: { componentName, property },
          dataDependencies,
        } = stateReference;
        return JSON.stringify({ reference: { componentName, property }, dataDependencies: dataDependencies.sort() });
      }),
    ),
  ];
  return dedupedSerializedRefs.map((ref) => JSON.parse(ref));
}

export function buildStateStatements(
  component: StudioComponent,
  componentMetadata: ComponentMetadata,
  importCollection: ImportCollection,
  apiKind: DataApiKind = 'DataStore',
): Statement[] {
  const stateHook = apiKind === 'DataStore' ? ImportValue.USE_STATE_MUTATION_ACTION : ImportValue.USE_STATE;
  if (componentMetadata.stateReferences.length > 0) {
    importCollection.addMappedImport(stateHook);
  }
  const dedupedStateReferences = dedupeStateReferences(componentMetadata.stateReferences);
  return dedupedStateReferences.map((stateReference) => {
    return factory.createVariableStatement(
      undefined,
      factory.createVariableDeclarationList(
        [
          factory.createVariableDeclaration(
            factory.createArrayBindingPattern([
              factory.createBindingElement(
                undefined,
                undefined,
                factory.createIdentifier(getStateName(stateReference.reference)),
                undefined,
              ),
              factory.createBindingElement(
                undefined,
                undefined,
                factory.createIdentifier(getSetStateName(stateReference.reference)),
                undefined,
              ),
            ]),
            undefined,
            undefined,
            factory.createCallExpression(factory.createIdentifier(stateHook), undefined, [
              getStateInitialValue(component, componentMetadata, stateReference),
            ]),
          ),
        ],
        ts.NodeFlags.Const,
      ),
    );
  });
}

export function getStateInitialValue(
  component: StudioComponent,
  componentMetadata: ComponentMetadata,
  stateReference: StateReferenceMetadata,
) {
  const {
    reference: { componentName, property },
    dataDependencies,
  } = stateReference;
  const referencedComponent = getComponentFromComponentTree(component, componentName);
  const componentProperty = referencedComponent.properties[property];

  if (componentProperty === undefined || dataDependencies.length > 0) {
    const defaultPropMapping = PrimitiveDefaultValuePropMapping[referencedComponent.componentType as Primitive];
    if (property in defaultPropMapping && referencedComponent.properties[defaultPropMapping[property]]) {
      const defaultProp = referencedComponent.properties[defaultPropMapping[property]];
      // eslint-disable-next-line no-param-reassign
      delete referencedComponent.properties[defaultPropMapping[property]];
      return propertyToExpression(componentMetadata, defaultProp);
    }

    return propertyToExpression(componentMetadata, getDefaultForComponentAndProperty(referencedComponent, property));
  }

  return propertyToExpression(componentMetadata, componentProperty);
}

export function getDefaultForComponentAndProperty(
  component: StudioComponent | StudioComponentChild,
  property: string,
): StudioComponentProperty | undefined {
  const { componentType } = component;
  const componentDefault = PrimitiveDefaultPropertyValue[componentType as Primitive];
  if (componentDefault && property in componentDefault) {
    // if component has defaultValue use defaultValue as initial state value
    return componentDefault[property];
  }

  // use empty string as fallback default
  return undefined;
}

export type MutationReferences = {
  [property: string]: { addControlEvent: boolean }[];
};

export function filterStateReferencesForComponent(
  component: StudioComponent | StudioComponentChild,
  stateReferences: StateReferenceMetadata[],
): MutationReferences {
  return stateReferences
    .filter(({ reference: { componentName } }) => componentName === component.name)
    .reduce(mutationReferenceReducerWithComponentType(component.componentType), {});
}

function mutationReferenceReducerWithComponentType(componentType: string) {
  return function mutationReferenceReducer(
    mutationReferences: MutationReferences,
    stateReference: StateReferenceMetadata,
  ): MutationReferences {
    const { reference } = stateReference;
    const propertyReferences = reference.property in mutationReferences ? mutationReferences[reference.property] : [];
    return {
      ...mutationReferences,
      [reference.property]: propertyReferences.concat([
        { addControlEvent: PrimitivesWithChangeEvent.has(componentType as Primitive) },
      ]),
    };
  };
}
