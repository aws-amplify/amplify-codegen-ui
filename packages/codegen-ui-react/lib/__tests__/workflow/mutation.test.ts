/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { MutationAction, DataStoreUpdateItemAction } from '@aws-amplify/codegen-ui';
import { getComponentStateReferences, getActionStateParameters } from '../../workflow/mutation';

describe('getComponentStateReferences', () => {
  test('basic', () => {
    const clickEvent: DataStoreUpdateItemAction = {
      action: 'Amplify.DataStoreUpdateItemAction',
      parameters: {
        model: 'Customer',
        id: {
          value: 'd9887268-47dd-4899-9568-db5809218751',
        },
        fields: {
          username: {
            componentName: 'UserNameTextField',
            property: 'value',
          },
        },
      },
    };

    const component = {
      id: '1234-5678-9010',
      componentType: 'Flex',
      name: 'MyForm',
      properties: {},
      bindingProperties: {},
      children: [
        {
          componentType: 'TextField',
          name: 'UsernameTextField',
          properties: {
            label: {
              value: 'Username',
            },
            value: {
              value: 'vizsla',
            },
          },
          bindingProperties: {},
        },
        {
          componentType: 'Button',
          name: 'SubmitButton',
          properties: {
            label: {
              value: 'Username',
            },
            value: {
              value: 'vizsla',
            },
          },
          bindingProperties: {},
          events: {
            click: clickEvent,
          },
        },
      ],
    };
    expect(getComponentStateReferences(component)).toMatchSnapshot();
  });
});

describe('getActionStateParameters', () => {
  test('basic', () => {
    const action: MutationAction = {
      action: 'Amplify.Mutation',
      parameters: {
        state: {
          componentName: 'ColoredBox',
          property: 'backgroundColor',
          set: {
            value: 'something',
          },
        },
      },
    };
    expect(getActionStateParameters(action)).toMatchSnapshot();
  });
});
