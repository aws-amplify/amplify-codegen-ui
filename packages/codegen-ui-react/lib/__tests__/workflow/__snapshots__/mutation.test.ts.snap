// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getActionStateParameters basic 1`] = `
Array [
  Object {
    "componentName": "ColoredBox",
    "property": "backgroundColor",
    "set": Object {
      "value": "something",
    },
  },
]
`;

exports[`mapSyntheticStateReferences basic 1`] = `
Array [
  Object {
    "componentName": "UserNameTextField",
    "property": "value",
  },
]
`;
